File: src/player/PlayerCommand.py
"""
PlayerCommand module: contains the PlayerCommand class

This class contains the inputs of a player for a particular server tick.  The
server runs the commands and the client predicts his own commands.
"""

from panda3d.core import Vec3, Vec2

from .InputButtons import InputFlag

import random

class PlayerCommand:

    def __init__(self, init=True):
        if init:
            self.clear()

    def makeCopy(self):
        p = PlayerCommand(False)
        p.buttons = self.buttons
        p.viewAngles = Vec3(self.viewAngles)
        p.mouseDelta = Vec2(self.mouseDelta)
        p.move = Vec3(self.move)
        p.weaponSelect = self.weaponSelect
        p.hasBeenPredicted = self.hasBeenPredicted
        p.tickCount = self.tickCount
        p.commandNumber = self.commandNumber
        p.randomSeed = self.randomSeed
        return p

    def clear(self):
        # Gameplay button states
        self.buttons = InputFlag.Empty

        # Player view angles
        self.viewAngles = Vec3()

        self.mouseDelta = Vec2()

        self.move = Vec3()

        self.weaponSelect = -1

        self.hasBeenPredicted = False

        self.tickCount = 0
        self.commandNumber = 0
        self.randomSeed = 0

    @staticmethod
    def readDatagram(dgi, prev):
        # Assume no change.
        cmd = prev.makeCopy()

        if dgi.getUint8():
            cmd.commandNumber = dgi.getUint32()
        else:
            # Assume steady increment.
            cmd.commandNumber = prev.commandNumber + 1

        rand = random.Random(cmd.commandNumber)
        cmd.randomSeed = rand.randint(0, 0xFFFFFFFF)

        if dgi.getUint8():
            cmd.tickCount = dgi.getUint32()
        else:
            # Assume steady increment.
            cmd.tickCount = prev.tickCount + 1

        if dgi.getUint8():
            cmd.viewAngles.readDatagramFixed(dgi)

        if dgi.getUint8():
            cmd.move.readDatagramFixed(dgi)

        if dgi.getUint8():
            cmd.buttons = dgi.getUint32()

        if dgi.getUint8():
            cmd.weaponSelect = dgi.getInt8()

        if dgi.getUint8():
            cmd.mouseDelta.readDatagramFixed(dgi)

        return cmd

    def writeDatagram(self, dg, prev):
        if self.commandNumber != (prev.commandNumber + 1):
            dg.addUint8(1)
            dg.addUint32(self.commandNumber)
        else:
            dg.addUint8(0)

        if self.tickCount != (prev.tickCount + 1):
            dg.addUint8(1)
            dg.addUint32(self.tickCount)
        else:
            dg.addUint8(0)

        if self.viewAngles != prev.viewAngles:
            dg.addUint8(1)
            self.viewAngles.writeDatagramFixed(dg)
        else:
            dg.addUint8(0)

        if self.move != prev.move:
            dg.addUint8(1)
            self.move.writeDatagramFixed(dg)
        else:
            dg.addUint8(0)

        if self.buttons != prev.buttons:
            dg.addUint8(1)
            dg.addUint32(self.buttons)
        else:
            dg.addUint8(0)

        if self.weaponSelect != prev.weaponSelect:
            dg.addUint8(1)
            dg.addInt8(self.weaponSelect)
        else:
            dg.addUint8(0)

        if self.mouseDelta != prev.mouseDelta:
            dg.addUint8(1)
            self.mouseDelta.writeDatagramFixed(dg)
        else:
            dg.addUint8(0)

File: src/cxxapp/playerCommand.cxx
#include "playerCommand.h"
#include "randomizer.h"
#include <limits>

NetworkClass *PlayerCommandArgs::_network_class = nullptr;
TypeHandle PlayerCommand::_type_handle;

/**
 *
 */
void PlayerCommand::
clear() {
  buttons = 0;
  view_angles = 0.0f;
  mouse_delta = 0.0f;
  move = 0.0f;
  weapon_select = -1;
  has_been_predicted = false;
  tick_count = 0;
  command_number = 0;
  random_seed = 0;
}

/**
 *
 */
void PlayerCommand::
read(DatagramIterator &scan, const PlayerCommand &prev) {
  if (scan.get_bool()) {
    command_number = scan.get_uint32();
  } else {
    // Assume steady increment.
    command_number = prev.command_number + 1;
  }

  Randomizer random(command_number);
  random_seed = random.random_int(std::numeric_limits<int>::max());

  if (scan.get_bool()) {
    tick_count = scan.get_uint32();
  } else {
    // Assume steady increment.
    tick_count = prev.tick_count + 1;
  }

  if (scan.get_bool()) {
    view_angles.read_datagram_fixed(scan);
  }

  if (scan.get_bool()) {
    move.read_datagram_fixed(scan);
  }

  if (scan.get_bool()) {
    buttons = scan.get_uint32();
  }

  if (scan.get_bool()) {
    weapon_select = scan.get_int8();
  }

  if (scan.get_bool()) {
    mouse_delta.read_datagram_fixed(scan);
  }
}

/**
 *
 */
void PlayerCommand::
write(Datagram &dg, const PlayerCommand &prev) {
  if (command_number != (prev.command_number + 1)) {
    dg.add_bool(true);
    dg.add_uint32(command_number);
  } else {
    dg.add_bool(false);
  }

  if (tick_count != (prev.tick_count + 1)) {
    dg.add_bool(true);
    dg.add_uint32(tick_count);
  } else {
    dg.add_bool(false);
  }

  if (view_angles != prev.view_angles) {
    dg.add_bool(true);
    view_angles.write_datagram_fixed(dg);
  } else {
    dg.add_bool(false);
  }

  if (move != prev.move) {
    dg.add_bool(true);
    move.write_datagram_fixed(dg);
  } else {
    dg.add_bool(false);
  }

  if (buttons != prev.buttons) {
    dg.add_bool(true);
    dg.add_uint32(buttons);
  } else {
    dg.add_bool(false);
  }

  if (weapon_select != prev.weapon_select) {
    dg.add_bool(true);
    dg.add_int8(weapon_select);
  } else {
    dg.add_bool(false);
  }

  if (mouse_delta != prev.mouse_delta) {
    dg.add_bool(true);
    mouse_delta.write_datagram_fixed(dg);
  } else {
    dg.add_bool(false);
  }
}
